store/tikv/store_aliveness.go

```diff
@@ -0,0 +1,273 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikv
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/pingcap/failpoint"
+	"net/http"
+	"strconv"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/pdapi"
+	"go.uber.org/zap"
+	"golang.org/x/sync/singleflight"
+)
+
+const knobSwitch = "livenessKnob"
+
+// StoreLiveness maintains the liveness of store.
+type StoreLiveness struct {
+	sf             singleflight.Group
+	timeout        time.Duration
+	statusResolver *kvStatusAddrResolver
+	doneCh         chan struct{}
+	knobs          struct {
+		mockTiKVStatus func(statusAddr string) error
+	}
+}
+
+func newStoreLiveness(pdHosts []string, timeout time.Duration) *StoreLiveness {
+	if len(pdHosts) == 0 {
+		return nil
+	}
+	doneCh := make(chan struct{})
+	l := &StoreLiveness{
+		timeout:        timeout,
+		statusResolver: newStatusAddressResolver(pdHosts, doneCh),
+		doneCh:         doneCh,
+	}
+	return l
+}
+
+// IsAlive checks whether store process still alive.
+func (s *StoreLiveness) IsAlive(storeID uint64) bool {
+	if s == nil {
+		return false
+	}
+	_, err, _ := s.sf.Do(strconv.FormatUint(storeID, 10), func() (interface{}, error) {
+		err := s.checkKvStatusAPI(storeID)
+		if err != nil {
+			logutil.BgLogger().Warn("[liveness] send fail and store is down, need invalid region cache and switch next peer",
+				zap.Uint64("store", storeID), zap.Error(err))
+			return nil, err
+		}
+		return nil, nil
+	})
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+func (s *StoreLiveness) start() error {
+	failpoint.Inject(knobSwitch, func() {
+		if s.statusResolver.knobs.mockStart != nil {
+			failpoint.Return(s.statusResolver.knobs.mockStart(s.statusResolver))
+		}
+	})
+	return s.statusResolver.start()
+}
+
+func (s *StoreLiveness) close() {
+	if s == nil {
+		return
+	}
+	close(s.doneCh)
+}
+
+func (s *StoreLiveness) checkKvStatusAPI(storeID uint64) error {
+	statusAddr := s.statusResolver.resolve(storeID)
+	if len(statusAddr) == 0 {
+		return errors.Errorf("unknown store %d maybe caused by offline in pd or cache outdated", storeID)
+	}
+	failpoint.Inject(knobSwitch, func() {
+		if s.knobs.mockTiKVStatus != nil {
+			failpoint.Return(s.knobs.mockTiKVStatus(statusAddr))
+		}
+	})
+	ctx, cancel := context.WithTimeout(context.Background(), s.timeout)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, statusAddr, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := util.InternalHTTPClient().Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err1 := resp.Body.Close()
+		if err1 != nil {
+			logutil.BgLogger().Error("close kv status api body failed", zap.Error(err))
+		}
+	}()
+	if resp.StatusCode != http.StatusOK {
```

Comment:
The server could run without exporting the status port? or hide it behind the cloud.

Reply:
good catch, I found my previous impl have question "when all pd down, tidb-server cannot start" question, fixed in last commit.

I prod env "status port" is accessible(we also have other function like 'cluster table' need it), tikv/kv/pd also configurated their status port with k8s's `--advertise-addr` option to make others can access it.

at last if "status port" is not accessible the store liveness can be degraded to works as current logic.

Reply:
use status api is @coocood 's advise, maybe the question is pd's kv status port info is only available in 4.0, it's hard to cherry-pick this pr to 3.0 or 3.1